### ISSUE TITLE
Refactor add-on for unit testing

### ISSUE THREAD

**Issue by lmorchard** [mozilla/testpilot#1220]
In PR #1218, I claim to fix #352. That's true, insofar as integration tests will exercise the add-on. Still, it would also be nice to have some good unit tests for the add-on by itself. 

However, I found it hard to write unit tests with all the dependencies and Firefox APIs the add-on uses. We have some magical tools for injecting mocks in Python & Node, but I couldn't find something like that for Firefox add-ons.

So, to get good unit test coverage, I think significant clean up & refactoring is necessary:
1. Finish #610 and everything we can from `index.js` into submodules.
2. Rework submodules to use [a dependency injection pattern](http://jasonpolites.github.io/tao-of-testing/ch3-1.1.html): implement constructors / initializers to be called from `index.js` that supply the submodule with any `require()`'d modules or instances of Firefox services.
3. Write unit tests that swap in mock APIs & services for the submodules' constructors & initializers, so they work against test code.

This seems kind of drastic, so I'd really like it if someone came along and told me there was a better way :)

**Comment by lmorchard** [comment]
A pointer from addons in IRC: 

https://developer.mozilla.org/en-US/Add-ons/SDK/Low-Level_APIs/_loader

tl;dr: Might be able to write a custom loader so that mock modules get loaded during testing. Need to investigate.

**Comment by jaredhirsch** [comment]
I'm happy to review when is ready

**Comment by lmorchard** [comment]
Hoping to have this in a PR on Monday. Two main things left:
- [ ] Finish 5 more tests to complete the suite
- [ ] Get this stuff running in Circle CI
